drivers/interrupt_controller/uwp_ictl.c

```diff
@@ -0,0 +1,207 @@
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	uwp_intc_disable(intc, irq);
+}
+
+static inline u32_t uwp_ictl_irq_get_state(struct device *dev)
+{
+	volatile struct uwp_intc *intc = INTC_STRUCT(dev);
+
+	return uwp_intc_status(intc);
+}
+
+static const struct irq_next_level_api uwp_ictl_apis = {
+	.intr_enable = uwp_ictl_irq_enable,
+	.intr_disable = uwp_ictl_irq_disable,
+	.intr_get_state = uwp_ictl_irq_get_state,
+};
+
+#ifdef CONFIG_UWP_ICTL_0
+static void uwp_config_0_irq(struct device *dev);
+static struct device DEVICE_NAME_GET(uwp_ictl_0);
+static struct uwp_ictl_config uwp_ictl_0_config = {
+	.irq_num = CONFIG_UWP_ICTL_0_IRQ,
+	.isr_table_offset = CONFIG_UWP_ICTL_0_OFFSET,
+	.config_func = uwp_config_0_irq,
+};
+
+static struct uwp_ictl_data uwp_ictl_0_data = {
+	.base_addr = CONFIG_UWP_ICTL_0_BASE,
+};
+
+static void uwp_config_0_irq(struct device *dev)
+{
+	IRQ_CONNECT(CONFIG_UWP_ICTL_0_IRQ,
+				CONFIG_UWP_ICTL_0_IRQ_PRIO,
+				uwp_ictl_isr,
+				DEVICE_GET(uwp_ictl_0), 0);
+	irq_enable(CONFIG_UWP_ICTL_0_IRQ);
+}
+
+static int uwp_ictl_0_init(struct device *dev)
+{
+	struct uwp_ictl_config *config = DEV_CFG(dev);
+
+	uwp_sys_enable(BIT(APB_EB_INTC));
+	uwp_sys_reset(BIT(APB_EB_INTC));
+
+	config->config_func(dev);
+
+	return 0;
+}
+
+DEVICE_AND_API_INIT(uwp_ictl_0, CONFIG_UWP_ICTL_0_NAME,
+		    uwp_ictl_0_init, &uwp_ictl_0_data, &uwp_ictl_0_config,
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+			&uwp_ictl_apis);
+#endif /* CONFIG_UWP_ICTL_0 */
+
+#ifdef CONFIG_UWP_ICTL_1
+static void uwp_config_1_irq(struct device *dev);
+static struct device DEVICE_NAME_GET(uwp_ictl_1);
+static struct uwp_ictl_config uwp_ictl_1_config = {
+	.irq_num = CONFIG_UWP_ICTL_1_IRQ,
+	.isr_table_offset = CONFIG_UWP_ICTL_1_OFFSET,
+	.config_func = uwp_config_1_irq,
+};
+
+static struct uwp_ictl_data uwp_ictl_1_data = {
+	.base_addr = CONFIG_UWP_ICTL_0_BASE + sizeof(struct uwp_intc),
+};
+
+static void uwp_config_1_irq(struct device *dev)
+{
+	IRQ_CONNECT(CONFIG_UWP_ICTL_1_IRQ,
+				CONFIG_UWP_ICTL_1_IRQ_PRIO,
+				uwp_ictl_isr,
+				DEVICE_GET(uwp_ictl_1), 0);
+	irq_enable(CONFIG_UWP_ICTL_1_IRQ);
+}
+
+static int uwp_ictl_1_init(struct device *dev)
+{
+	struct uwp_ictl_config *config = DEV_CFG(dev);
+
+	uwp_sys_enable(BIT(APB_EB_INTC));
+	uwp_sys_reset(BIT(APB_EB_INTC));
+
+	config->config_func(dev);
+
+	return 0;
+}
+
+DEVICE_AND_API_INIT(uwp_ictl_1, CONFIG_UWP_ICTL_1_NAME,
+		    uwp_ictl_1_init, &uwp_ictl_1_data, &uwp_ictl_1_config,
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+			&uwp_ictl_apis);
+#endif /* CONFIG_UWP_ICTL_1 */
+
+#ifdef CONFIG_UWP_ICTL_2
+static void uwp_config_2_irq(struct device *dev);
+static struct device DEVICE_NAME_GET(uwp_ictl_2);
+static struct uwp_ictl_config uwp_ictl_2_config = {
+	.irq_num = CONFIG_UWP_ICTL_2_IRQ,
+	.isr_table_offset = CONFIG_UWP_ICTL_2_OFFSET,
+	.config_func = uwp_config_2_irq,
+};
+
+static struct uwp_ictl_data uwp_ictl_2_data = {
+	.base_addr = CONFIG_UWP_ICTL_2_BASE,
```

Comment:
Can't this be defined similar to what was done for ictl_1? i.e. .base_addr = CONFIG_UWP_ICTL_0_BASE + sizeof(struct uwp_intc)

Reply:
This is because ICTL_0 and ICTL_1 are combined in one interrupt controller, and ICTL_2 is another one. Therefore, the base address of ICTL_2 is not related with ICTL_0 and ICTL_1.